File: CFOnlineShop/CFOnlineShop/ViewController/CFShoppingCartController/CFShoppingCartController.m
//
//  CFShoppingCartController.m
//  CFOnlineShop
//
//  Created by chenfeng on 2018/7/18.
//  Copyright © 2018年 chenfeng. All rights reserved.
//

#import "CFShoppingCartController.h"
#import "CFShoppingCartCell1.h"
#import "CFShoppingCartCell2.h"
#import "CFShoppingCartHeaderView.h"

@interface CFShoppingCartController ()<UICollectionViewDelegate,UICollectionViewDataSource>

@property (nonatomic, strong) UICollectionView *collectionView;

@end

@implementation CFShoppingCartController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    [self setTitle:@"购物车"];
    self.navigationView.backgroundColor = kWhiteColor;
    [self setCollectionView];
    
}

- (void)setCollectionView
{
    UICollectionViewFlowLayout *flowLayout = [[UICollectionViewFlowLayout alloc] init];
    _collectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, TopHeight, Main_Screen_Width, Main_Screen_Height - TabbarHeight - TopHeight) collectionViewLayout:flowLayout];
    _collectionView.dataSource = self;
    _collectionView.delegate = self;
    _collectionView.backgroundColor = kClearColor;
    [_collectionView registerClass:[CFShoppingCartCell1 class] forCellWithReuseIdentifier:@"CollectionCell"];
    [_collectionView registerClass:[CFShoppingCartCell2 class] forCellWithReuseIdentifier:@"CollectionCell2"];
    [_collectionView registerClass:[CFShoppingCartHeaderView class] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:@"header"];
    [self.view addSubview:_collectionView];
    
    //去掉顶部偏移
    if (@available(iOS 11.0, *))
    {
        _collectionView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
    }
    
}

#pragma mark - UICollectionViewDataSource

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView{
    return 2;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    if (section == 0) {
        return 5;
    }
    return 10;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    
    if (indexPath.section == 0) {
        
        static NSString *collectionCell = @"CollectionCell";
        CFShoppingCartCell1 *cell = [collectionView dequeueReusableCellWithReuseIdentifier:collectionCell forIndexPath:indexPath];
        [cell configCollectionCellType:(CFEditCollectionCellTypeWithDelete)];
        
        //cell.backgroundColor = kRedColor;
        cell.titleStr.text = @"测试商品";
        NSString *imageName = [NSString stringWithFormat:@"commodity_%ld",(long)indexPath.row + 1];
        cell.imageView.image = [UIImage imageNamed:imageName];
        [cell setDeleteButtonAction:^(UIButton *button) {
            NSLog(@"删除操作");
        }];
        
        return cell;
    }
    else
    {
        static NSString *collectionCell2 = @"CollectionCell2";
        CFShoppingCartCell2 *cell = [collectionView dequeueReusableCellWithReuseIdentifier:collectionCell2 forIndexPath:indexPath];
        [cell configCollectionCellType:(CFEditCollectionCellTypeWithNone)];
        cell.backgroundColor = kRedColor;
        cell.titleStr.text = @"测试商品";
        
        NSString *imageName = [NSString stringWithFormat:@"commodity_%ld",(long)indexPath.row + 1];
        
        cell.imageView.image = [UIImage imageNamed:imageName];
        [cell setDeleteButtonAction:^(UIButton *button) {
            NSLog(@"删除操作");
        }];
        
        return cell;
    }
}


#pragma mark -- UICollectionViewDelegateFlowLayout

- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section
{
    return 5;
}

- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section{
    
    return 5;
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.section == 0) {
        
        return CGSizeMake(_collectionView.mj_w - 10, 80);
    }
    else
    {
        NSInteger itemCount = 2;
        return CGSizeMake((Main_Screen_Width - 25)/itemCount, (Main_Screen_Width - 25)/itemCount + 60);
    }
}

//距离collectionview的上下左右边距
- (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section
{
    return UIEdgeInsetsMake(5,10,0,10);
}

- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath{
    
    UICollectionReusableView *reusableview = nil;
    
    if (kind == UICollectionElementKindSectionHeader && indexPath.section == 1){
        
        CFShoppingCartHeaderView *headerView = [collectionView dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:@"header" forIndexPath:indexPath];
        reusableview = headerView;
        //reusableview.backgroundColor = kRedColor;
    }
    
    return reusableview;
}


- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout referenceSizeForHeaderInSection:(NSInteger)section
{
    if (section == 0) {
        return CGSizeMake(0, 0);
    }
    
    return CGSizeMake(Main_Screen_Width, 40);
}

#pragma mark --UICollectionViewDelegate

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    NSLog(@"click collectionView row");
}

#pragma mark --UIScrollViewDelegate

- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
{
    NSLog(@"开始拖动");
    [[_collectionView visibleCells] makeObjectsPerformSelector:@selector(hiddenButtonsWithAnimation)];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: SwiftOnlineShopDemo/SwiftOnlineShopDemo/ViewController/CFShoppingCartController/CFShoppingCartController.swift
//
//  CFShoppingCartController.swift
//  SwiftOnlineShopDemo
//
//  Created by chenfeng on 2019/4/23.
//  Copyright © 2019年 chenfeng. All rights reserved.
//

import UIKit

var currentCellForKey = "currentCell"

//extension UICollectionView{
//
//    var currentCell: UICollectionViewCell?{
//        set {
//            objc_setAssociatedObject(self, &keyName, newValue, objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC)
//        }
//
//        get {
//            if let rs = objc_getAssociatedObject(self, &keyName) as? UICollectionViewCell {
//                return rs
//            }
//            return nil
//        }
//    }
//
//}

class CFShoppingCartController: CFBaseController,UICollectionViewDelegate,UICollectionViewDataSource,UICollectionViewDelegateFlowLayout {

    var collectionView: UICollectionView!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.navigationView.backgroundColor = UIColor.white
        self.view.backgroundColor = BACKGROUND_VIEW_COLOR
        self.setTitle(title: "购物车")
        setCollectionView()
        // Do any additional setup after loading the view.
    }
    
    func setCollectionView() {
        
        let layout = UICollectionViewFlowLayout.init()
        
        collectionView = UICollectionView.init(frame: CGRect(x:0, y:STATUS_AND_NAV_BAR_HEIGHT, width:SCREEN_WIDTH, height:SCREEN_HEIGHT - TABBAR_HEIGHT - STATUS_AND_NAV_BAR_HEIGHT), collectionViewLayout: layout)
        collectionView.backgroundColor = UIColor.clear
        collectionView.delegate = self
        collectionView.dataSource = self
        collectionView.register(CFShoppingCartCell1.classForCoder(), forCellWithReuseIdentifier: "CollectionCell")
        collectionView.register(CFShoppingCartCell2.classForCoder(), forCellWithReuseIdentifier: "CollectionCell2")
        collectionView.register(CFShoppingCartHeaderView.classForCoder(), forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "header")
        collectionView.showsVerticalScrollIndicator = false
        self.view.addSubview(collectionView)
        
        //去掉顶部偏移
        if #available(iOS 11.0, *)
        {
            collectionView.contentInsetAdjustmentBehavior = .never;
        }
        
    }

    // MARK: - UICollectionView Method
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 2
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        if section == 0 {
            return 5
        }
        return 10
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        
        if indexPath.section == 0 {
            let collectionCellIden = "CollectionCell";

            let cell:CFShoppingCartCell1! = collectionView.dequeueReusableCell(withReuseIdentifier: collectionCellIden, for: indexPath) as? CFShoppingCartCell1
            
            cell.configCollection(cellType: .deleting)
            
            cell.titleStr.text = "测试商品"
            
            let imageName = String(format: "%@%d","commodity_",indexPath.row + 1)
            
            cell.imageView.image = UIImage(named: imageName)
            
            cell.deleteButtonAction = {(button:UIButton) in
                print("删除操作")
            }
            
            return cell
        }
        else
        {
            let collectionCellIden = "CollectionCell2";
            
            let cell:CFShoppingCartCell2! = collectionView.dequeueReusableCell(withReuseIdentifier: collectionCellIden, for: indexPath) as? CFShoppingCartCell2
            
            cell.configCollection(cellType: .none)
            
            cell.titleStr.text = "测试商品"
            
            let imageName = String(format: "%@%d","commodity_",indexPath.row + 1)
            
            cell.imageView.image = UIImage(named: imageName)
            
            cell.deleteButtonAction = {(button:UIButton) in
                print("删除操作")
            }
            
            return cell
        }
        
    }
    
    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        
        var reusableview : UICollectionReusableView?
        
        if kind == UICollectionView.elementKindSectionHeader && indexPath.section == 1{
            
            let headerView:CFShoppingCartHeaderView! = collectionView.dequeueReusableSupplementaryView(ofKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "header", for: indexPath) as? CFShoppingCartHeaderView
            
            reusableview = headerView;
        }
        
        return reusableview ?? UIView() as! UICollectionReusableView
        
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForHeaderInSection section: Int) -> CGSize {
        
        if section == 0 {
            return CGSize(width: 0, height: 0)
        }
        
        return CGSize(width: SCREEN_WIDTH, height: 40)
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
        return 5
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
        return 5
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        
        if indexPath.section == 0 {
            return CGSize(width: collectionView.mj_w - 10, height: 80)
        }
        
        let count:CGFloat = 2
        return CGSize(width: (SCREEN_WIDTH - 25)/count, height: (SCREEN_WIDTH - 25)/count + 60)
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
        return UIEdgeInsets(top: 5, left: 10, bottom: 0, right: 10);
    }
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        
        
    }
    
    
    // MARK: - UIScrollViewDelegate
    func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        print("开始拖动");
        collectionView.visibleCells.forEach { (cell) in
            (cell as! CFEditCollectionCell).hiddenButtonsWithAnimation()
        }
    }
    
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}
